game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/BattleStep.java

```diff
@@ -0,0 +1,88 @@
+package games.strategy.triplea.delegate.battle.steps;
+
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.IExecutable;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import java.util.List;
+import lombok.AllArgsConstructor;
+
+/**
+ * A step in a battle.
+ *
+ * <ol>
+ *   <ul>
+ *     Each step can have 0 or more names. These names are shown in the {@link
+ *     games.strategy.triplea.ui.BattlePanel}
+ *   </ul>
+ *   <ul>
+ *     Each step can also have an executable. See {@link BattleAtomic} on what an executable is
+ *     comprised of.
+ *   </ul>
+ * </ol>
+ */
+@AllArgsConstructor
+public abstract class BattleStep {
+
+  /**
+   * The current state of the battle
+   *
+   * <p>Note: This is current mutable so the underlying data can change.
+   */
+  protected final BattleState battleState;
+
+  /** Actions that can occur in a battle that require interaction with {@link IDelegateBridge} */
+  protected final BattleActions battleActions;
+
+  public abstract BattleAtomic getExecutable();
+
+  public abstract List<String> getNames();
+
+  /**
+   * Determine if this step should run based on the request
+   *
+   * @return true if valid
+   */
+  public abstract boolean valid();
+
+  /**
+   * Executes the step
+   *
+   * <p>This is called by the BattleAtomic and {@link #valid} has already been checked
```

Comment:
Typically methods should not be aware of who is calling them. It's also a design smell when we require one method to have been called before another (and furthermore when that method is determining state that dictates whether you can and should call another method).

That could potentially be fixed by returning an object to then be executed. EG:

```
protected Optional<BiConsumer<ExecutionStack, IDelegateBridge>> getBattleStepExecutable();
```

Then usage below of:
```
 public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
      final BattleStep executingStep = BattleStep.this;
      if (executingStep.valid()) {
        executingStep.execute(stack, bridge);
      }
```

would be transformed into:
```
 public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
       getBattleStepExecutable().ifPresent(stepExecutable -> stepExecutable.accept(stack, bridge));
  }
```

It also appears that `getName()` has a dependency on `isValid()` as well. In those cases if `isValid()` is false then presumably `getNames()` could return an empty collection.

WDYT?

Reply:
`getBattleStepExecutable` actually returns the `BattleAtomic`.  So calling `getBattleStepExecutable` inside of the `BattleAtomic` will not work since it will just go into an infinite loop.

I see what you are trying to do but I don't see how that helps.  You are just moving the `valid()` call to another method which is still making `execute()` be dependent on `valid()`.  I had made `execute()` a protected method so that the dependency on `valid()` was controlled by the class and callers wouldn't have to worry about it.  It was an internal implementation.  Though, with your idea of turning this class into an interface, that internal implementation becomes public, so it complicates it.  I guess implementations of BattleStep will just have to ensure that `execute()` checks `isValid()` beforehand.

As for `getNames()`, I can see why you say it has a dependency.  The only caller of it is always checking `isValid()`.  I can envision other callers not caring if it is valid or not but I currently don't envision creating those callers so I can move `isValid()` inside of each of the `getNames()`.